ngraph/core/reference/include/ngraph/runtime/reference/gather.hpp

```diff
@@ -27,149 +30,386 @@ namespace ngraph
+                    {
+                        params_outer_end_corner[i] = 1;
+                    }
+                    Strides params_outer_strides(params_ndim, 1);
+                    AxisVector params_outer_axis_order(params_ndim);
+                    std::iota(params_outer_axis_order.begin(), params_outer_axis_order.end(), 0);
+                    CoordinateTransform params_outer_transform(params_shape,
+                                                               params_outer_start_corner,
+                                                               params_outer_end_corner,
+                                                               params_outer_strides,
+                                                               params_outer_axis_order);
+
+                    // Create a CoordinateTransform for "indices" that visits only the first element
+                    // along inner most axis
+                    Coordinate indices_outer_start_corner(indices_ndim, 0);
+                    Coordinate indices_outer_end_corner(indices_shape);
+                    if (indices_ndim > 0)
+                    {
+                        indices_outer_end_corner[indices_ndim - 1] = 1;
+                    }
+                    Strides indices_outer_strides(indices_ndim, 1);
+                    AxisVector indices_outer_axis_order(indices_ndim);
+                    std::iota(indices_outer_axis_order.begin(), indices_outer_axis_order.end(), 0);
+                    CoordinateTransform indices_outer_transform(indices_shape,
+                                                                indices_outer_start_corner,
+                                                                indices_outer_end_corner,
+                                                                indices_outer_strides,
+                                                                indices_outer_axis_order);
+
+                    // Create an inner CoordinateTransfrom for "out"
+                    size_t out_inner_ndim = out_ndim - axis;
+                    Shape out_inner_shape(out_shape);
+                    out_inner_shape.erase(out_inner_shape.begin(), out_inner_shape.begin() + axis);
+                    Coordinate out_inner_start_corner(out_inner_ndim, 0);
+                    Coordinate out_inner_end_corner(out_inner_shape);
+                    if (indices_ndim > 0)
+                    {
+                        out_inner_end_corner[indices_ndim - 1] = 1;
+                    }
+                    for (size_t i = indices_ndim; i < out_inner_ndim; i++)
+                    {
+                        out_inner_end_corner[i] = 1;
+                    }
+                    Strides out_inner_strides(out_inner_ndim, 1);
+                    AxisVector out_inner_axis_order(out_inner_ndim);
+                    std::iota(out_inner_axis_order.begin(), out_inner_axis_order.end(), 0);
+                    CoordinateTransform out_inner_transform(out_inner_shape,
```

Comment:
Can we avoid using CoordinateTransform and CoordinateTransform::index?

Reply:
`v0::gather` will be removed when all will be finish.